geo: validate geospatial objects

otan opened the pull request:
This commit validate geospatial objects properties before their
creation. LineStrings should have at least 2 points, Polygons at least
4 per LinearRing and must be closed.

Resolves #51074.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51256)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51256)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geo.go, line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51256#-MBsupxD2F-P79emxabf:-MBsupxD2F-P79emxabg:b-77dt5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/cddbc9abfe4cb00c1eb1a0cfcc36601cf8576656/pkg/geo/geo.go#L654)):*
> ```Go
> // validateGeomT validates the geom.T object across valid geom.T objects,
> // returning an error if it is invalid.
> func validateGeomT(t geom.T) error {
> ```

Will all inserts using a SQL statement go through this validation?
If yes, can you add a couple of logic test cases to confirm the end-to-end behavior?
How does this validation compare to PostGIS?
___
*[pkg/geo/geomfn/make_geometry_test.go, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51256#-MBswGOn6YCfPd5WaXvX:-MBswGOn6YCfPd5WaXvY:blbv1ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/cddbc9abfe4cb00c1eb1a0cfcc36601cf8576656/pkg/geo/geomfn/make_geometry_test.go#L236)):*
> ```Go
> 			geopb.DefaultGeometrySRID,
> 			errors.Newf("Polygon LinearRing must have at least 4 points, found 2 at position 2"),
> 		},
> ```

can you add a test with a GEOMETRYCOLLECTION containing a MULTIPOLYGON containing a bad POLYGON, to show the final error string after the multiple layers of error wrapping?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51256)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geo.go, line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51256#-MBsupxD2F-P79emxabf:-MBt5ZhN2CBtnBSQfgdd:bg3gpdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cddbc9abfe4cb00c1eb1a0cfcc36601cf8576656/pkg/geo/geo.go#L654)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Will all inserts using a SQL statement go through this validation?
If yes, can you add a couple of logic test cases to confirm the end-to-end behavior?
How does this validation compare to PostGIS?

</blockquote></details>

Everything will go through this before being a fully fledged Geometry or Geography type.
Validation in PostGIS is spotty (distributed over the place I believe since error messages are different), but checks these cases.
___
*[pkg/geo/geomfn/make_geometry_test.go, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51256#-MBswGOn6YCfPd5WaXvX:-MBt69WwEwdneegr9f3j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cddbc9abfe4cb00c1eb1a0cfcc36601cf8576656/pkg/geo/geomfn/make_geometry_test.go#L236)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add a test with a GEOMETRYCOLLECTION containing a MULTIPOLYGON containing a bad POLYGON, to show the final error string after the multiple layers of error wrapping?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
geo: validate geospatial objects

This commit validate geospatial objects properties before their
creation. LineStrings should have at least 2 points, Polygons at least
4 per LinearRing and must be closed.

Release note: None

otan commented:
bors r=sumeerbhola
